Categorical TOC model and src folder organization (#63)
rossyndicate/upper_clp_dss: modeling/toc/03_ross_only_categorical_TOC_model.Rmd

---
title: "Testing TOC preds"
author: "Sam Struthers-ROSSyndicate"
date: "2025-08-01"
output: html_document
---

> 

Reviewer: suggestions: add a quick purpose statement here

```{r}
# Read in all data to create scaling parameters
wide_data_file <- list.files(here("data","collated", "sensor"), pattern = "all_sensor_data_.*\\.parquet", full.names = TRUE)%>%
  tail(1)

# Use the complete sensor dataset for scaling parameters (assuming sensor has all necessary columns already)
complete_sensor_data <- read_parquet(wide_data_file) %>%
  apply_sensor_transformations_toc(., dt_col = "DT_hourly_round")%>%
  select(-DO, -Depth, -ORP, -pH) #remove unused sensors

```

## Read in matchup dataset

```{r}

matchup_dataset <- list.files(here("data","derived", "matchup_datasets"), pattern = "parsed_toc_sensor_clp_", full.names = TRUE)%>%
  tail(1)

# Read in all matches data
all_matches <- read_csv(file = matchup_dataset, show_col_types = F) %>%
  apply_sensor_transformations_toc(dt_col = "sensor_datetime")%>%
  #remove NAs
  filter(!is.na(TOC))%>%
  mutate(clean_sensor_datetime = parse_date_time(sensor_datetime, orders = c("%Y-%m-%d %H:%M:%S", "%m/%d/%y %H:%M","%Y-%m-%d" )))%>%
  mutate(sensor_datetime = force_tz(clean_sensor_datetime, tzone = 'MST'), 

>          month = lubridate::month(sensor_datetime),

Reviewer: ```suggestion
         month = month(sensor_datetime),
```

         #create categorical TOC
         TOC_cat  = cut(TOC, breaks = c(0, 2, 4, 8, 15), labels = c('low', 'medium', 'high',"extreme")),
         TOC_cat = as.numeric(as.factor(TOC_cat))-1
  )

```

## Define ID Columns, Target, and Features

```{r}
# ID columns to keep to use later on 
id_cols <- c("Date", "id","site", "DT_hourly_round", "sensor_datetime","lab_datetime",  "collector", "data_avail")

> #What we want to predict

Reviewer: ```suggestion
# What we want to predict
```

  ) %>%
  na.omit()

```

## Create Holdout Dataset

For this model version, we will create a holdout testing set from specific sites (chd, sfm, pbr, pman, pbd) in 2025, and use all other data for training/validation. We will also trim out any known bad data or sites we don't want to include in modeling.

```{r}
# Trim out any known bad data or sites we don't want to include in modeling
all_matches_trimmed <- all_matches_norm%>%
  filter(!(id == "cbri" & between(sensor_datetime, ymd("2024-10-08"), ymd("2024-10-10"))))%>% #exclude a cbri sample taken during release
  filter(!(id == "pman" & between(sensor_datetime, ymd("2025-05-28"), ymd("2025-06-01"))))%>% #exclude a pman sample with bad sensor data
  #These are not applicable sites for modeling Upper CLP streams and will just introduce error
  filter( id %nin% c("springcreek", "boxcreek"))
# Create testing set from 2025 data at specific sites
testing <- all_matches_trimmed %>%
  filter((year(sensor_datetime) == 2025 & id %in% c("chd", "sfm", "pbr", "pman", "pbd")))
# Create training/validation set
train_val <- all_matches_trimmed %>%
  anti_join(testing)%>%
  select(id, any_of(features), TOC_cat, sensor_datetime, collector)%>%

>   mutate(id = ifelse(id == "archery_virridy", "archery", id)) #combine archery and archery_virridy since they are the same site

Reviewer: ```suggestion
  mutate(id = if_else(id == "archery_virridy", "archery", id)) #combine archery and archery_virridy since they are the same site
```

  bind_rows(
    testing %>%
      mutate(group = "Testing")
  ) %>%
  group_by(group, TOC_cat) %>%
  summarise(n = n(), .groups = "drop_last") %>%
  mutate(prop = n / sum(n))%>%
  mutate(
    TOC_cat_range = case_when(
      TOC_cat == 0 ~ "0-2",
      TOC_cat == 1 ~ "2-4",
      TOC_cat == 2 ~ "4-8",
      TOC_cat == 3 ~ "8+"
    ))

# Plot as heat map
ggplot(heat_data, aes(x = factor(TOC_cat_range), y = group, fill = prop)) +
  geom_tile(color = "white") +
  geom_text(aes(label = paste0(round(prop * 100, 1), "%")), color = "black", size = 3.5) +
  scale_fill_gradient(low = "#F5F5F5", high = "#256BF5") +
  labs(
    title = paste0(
      "Training/Validation Set: ", nrow(train_val), " samples              ",
      "Testing Set: ", nrow(testing), " samples"
    ),
    # subtitle = paste0(
    #   "Training TOC Range: ", round(min(train_val$TOC, na.rm = TRUE), 2), " - ",
    #   round(max(train_val$TOC, na.rm = TRUE), 2), " mg/L (Median: ", round(median(train_val$TOC), 2), ")\n",
    #   "Testing TOC Range: ", round(min(testing$TOC, na.rm = TRUE), 2), " - ",
    #   round(max(testing$TOC, na.rm = TRUE), 2), " mg/L (Median: ", round(median(testing$TOC), 2), ")"

>     # ),

Reviewer: ```suggestion
```

    #f_turb = c(general_features, "f_turb"), # FDOM / turbidity, high FDOM + low turb = high TOC (snowmelt in high elev reaches), redundant with f_x_turb
    #turb_f = c(general_features, "turb_f"), # turbidity / FDOM,  redundant with f_x_turb
    fewer_features = c(fewer_features), 
    fewer_features_trimmed = c(fewer_features)%>%setdiff(., c("turb_median_12hr","temp_x_f", "temp_mean_48hr","temp_x_turb", "turb_f", "sc_sd_24hr", "temp_mean_24hr", "temp_f"  )),
        fewest = c(fewer_features)%>%setdiff(., c("turb_median_12hr","temp_x_f", "temp_mean_48hr","temp_x_turb", "turb_f", "sc_sd_24hr", "temp_mean_24hr", "temp_f", "fdom_corr_max_6hr" ))
    
    #Trimming down to most important features from all features (based on importance and shap values) to reduce overfitting
    # trimmed_features = c(general_features, "fdom_corr_max_6hr", "f_sc", "f_x_turb", "sin_doy" , "turb_median_12hr", "daily_canyon_flow_cfs")
  ) , ~test_features(
    train_val = train_val,
    features_to_test = .x,
    target = "TOC_cat",
    default_hyper_params = default_hyper_params
  ))
```

> #### Feature Set Results

Reviewer: suggestion: can we reduce the number of TOC levels here? I think best practice is 3 or less.

Overall, this is a good sign that our tuning efforts have paid off!

## Save as "in_dev" and remove unneeded objects

If we want to test additional ideas for bias correction or look at model performance on timeseries data, we can just run this script until `model_training` and then load in the "in_dev" model file to avoid having to rerun all the tuning again.

```{r}
#commented out so we don't overwrite existing model files
# saveRDS(full_model, file = here("data", "derived", "models","TOC_realtime", "in_dev", paste0("toc_categorical_xgboost_models_",version,".rds")))
#if we want to reload model
#full_model <- readRDS(here("data", "derived", "models","TOC_realtime", "in_dev", paste0("toc_categorical_xgboost_models_",version,".rds")))

rm(list = setdiff(ls(), c("train_val", "testing", "full_model","version","id_cols",
                          "complete_sensor_data",
                          #keeping functions around
                          "apply_training_scale","plot_tvt_fold_acc","plot_train_test_ensemble_acc",

>                          "load_grab_sample_data", "plot_cat_model_timeseries" )) )

Reviewer: ```suggestion
                          "load_grab_sample_data", "plot_cat_model_timeseries")))
```
